File: AI Lab Assignments/8-Puzzle/Main.cpp
#include <iostream>
#include <vector>
#include <deque>
#include <set>
#include <algorithm>

using namespace std;

struct Node {
    vector<vector<int>> state;
    Node* parent;
    pair<int, int> move;
    int depth;

    Node(const vector<vector<int>>& s, Node* p, const pair<int, int>& m, int d) : state(s), parent(p), move(m), depth(d) {}
};

vector<Node> generate_successors(const Node& node) {
    vector<Node> successors;

    int blank_row = -1;
    int blank_col = -1;

    for (int i = 0; i < 3; ++i) {
        for (int j = 0; j < 3; ++j) {
            if (node.state[i][j] == 0) {
                blank_row = i;
                blank_col = j;
            }
        }
    }

    vector<pair<int, int>> moves = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

    for (const auto& move : moves) {
        int new_row = blank_row + move.first;
        int new_col = blank_col + move.second;

        if (new_row >= 0 && new_row < 3 && new_col >= 0 && new_col < 3) {
            vector<vector<int>> new_state = node.state;
            swap(new_state[blank_row][blank_col], new_state[new_row][new_col]);

            successors.emplace_back(new_state, &node, move, node.depth + 1);
        }
    }

    return successors;
}

bool operator==(const Node& lhs, const Node& rhs) {
    return lhs.state == rhs.state;
}

struct NodeHash {
    size_t operator()(const Node& node) const {
        size_t hash_value = 0;
        for (const auto& row : node.state) {
            for (int num : row) {
                hash_value = hash_value * 10 + num;
            }
        }
        return hash_value;
    }
};

vector<vector<Node>> solve_eight_puzzle(const vector<vector<int>>& start_state, const vector<vector<int>>& goal_state) {
    deque<Node> queue;
    set<Node, NodeHash> visited;
    vector<vector<Node>> solution_paths;

    Node root(start_state, nullptr, make_pair(0, 0), 0);
    queue.push_back(root);
    visited.insert(root);

    while (!queue.empty()) {
        Node current_node = queue.front();
        queue.pop_front();

        if (current_node.state == goal_state) {
            vector<Node> solution_path;
            Node* node_ptr = &current_node;
            while (node_ptr) {
                solution_path.insert(solution_path.begin(), *node_ptr);
                node_ptr = node_ptr->parent;
            }
            solution_paths.push_back(solution_path);
        }

        vector<Node> successors = generate_successors(current_node);

        for (const Node& successor : successors) {
            if (visited.find(successor) == visited.end()) {
                queue.push_back(successor);
                visited.insert(successor);
            }
        }
    }

    return solution_paths;
}

int main() {
    int n = 3;
    vector<vector<int>> start_state(n, vector<int>(n));
    vector<vector<int>> goal_state(n, vector<int>(n));

    cout << "Enter the start state:" << endl;
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            cin >> start_state[i][j];
        }
    }

    cout << "Enter the goal state:" << endl;
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            cin >> goal_state[i][j];
        }
    }

    vector<vector<Node>> both_solution_paths = solve_eight_puzzle(start_state, goal_state);

    int k = 0;
    for (const vector<Node>& solution_paths : both_solution_paths) {
        if (k == 0) {
            cout << "Solution Path:" << endl;
        } else {
            cout << "Alternative Solution Path:" << endl;
        }

        for (const Node& node : solution_paths) {
            for (const vector<int>& row : node.state) {
                for (int num : row) {
                    cout << num << " ";
                }
                cout << endl;
            }
            cout << endl;
        }

        ++k;
    }

    if (k == 0) {
        cout << "No Alternative Solution Exists" << endl;
    }

    return 0;
}

File: AI Lab Assignments/8-Puzzle/Main.js
class Node {
    constructor(state, parent, move, depth) {
        this.state = state;
        this.parent = parent;
        this.move = move;
        this.depth = depth;
    }
}

function generateSuccessors(node) {
    const successors = [];

    let blankRow = -1;
    let blankCol = -1;

    for (let i = 0; i < 3; i++) {
        for (let j = 0; j < 3; j++) {
            if (node.state[i][j] === 0) {
                blankRow = i;
                blankCol = j;
            }
        }
    }

    const moves = [[-1, 0], [1, 0], [0, -1], [0, 1]];

    for (const move of moves) {
        const newRow = blankRow + move[0];
        const newCol = blankCol + move[1];

        if (newRow >= 0 && newRow < 3 && newCol >= 0 && newCol < 3) {
            const newState = node.state.map(row => row.slice());
            newState[blankRow][blankCol] = newState[newRow][newCol];
            newState[newRow][newCol] = 0;

            successors.push(new Node(newState, node, move, node.depth + 1));
        }
    }

    return successors;
}

function solveEightPuzzle(startState, goalState) {
    const queue = [];
    const visited = new Set();
    const solutionPaths = [];

    const root = new Node(startState, null, null, 0);
    queue.push(root);
    visited.add(JSON.stringify(startState));

    while (queue.length > 0) {
        const currentNode = queue.shift();

        if (JSON.stringify(currentNode.state) === JSON.stringify(goalState)) {
            const solutionPath = [];
            let node = currentNode;
            while (node) {
                solutionPath.unshift(node);
                node = node.parent;
            }
            solutionPaths.push(solutionPath);
        }

        const successors = generateSuccessors(currentNode);

        for (const successor of successors) {
            const stateStr = JSON.stringify(successor.state);
            if (!visited.has(stateStr)) {
                queue.push(successor);
                visited.add(stateStr);
            }
        }
    }

    return solutionPaths;
}

function main() {
    const n = 3;
    const startState = [];
    const goalState = [];

    const readline = require('readline');
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });

    console.log("Enter the start state:");
    rl.on('line', line => {
        const rowValues = line.split(" ").map(Number);
        startState.push(rowValues);

        if (startState.length === n) {
            console.log("Enter the goal state:");
            rl.on('line', line => {
                const rowValues = line.split(" ").map(Number);
                goalState.push(rowValues);

                if (goalState.length === n) {
                    rl.close();
                    const bothSolutionPaths = findTwoSolutions(startState, goalState);
                    let k = 0;
                    for (const solutionPaths of bothSolutionPaths) {
                        if (k === 0) {
                            console.log("Solution Path:");
                        } else {
                            console.log("Alternative Solution Path:");
                        }
                        printSolutionPaths(solutionPaths);
                        if (bothSolutionPaths.length > 1) k++;
                    }
                    if (k === 0) console.log("No Alternative Solution Exists");
                }
            });
        }
    });
}

function findTwoSolutions(startState, goalState) {
    const bothSolutionPaths = [];

    // Find the first solution
    const solutionPath1 = solveEightPuzzle(startState, goalState);
    bothSolutionPaths.push(solutionPath1);

    // Find the second solution (alternative)
    // Implement this similar to solveEightPuzzle but change the order of moves

    return bothSolutionPaths;
}

function printSolutionPaths(solutionPaths) {
    if (solutionPaths.length === 0) {
        console.log("No solution found.");
        return;
    }

    for (const solutionPath of solutionPaths) {
        printSolutionPath(solutionPath);
        console.log();
    }
}

function printSolutionPath(solutionPath) {
    for (const node of solutionPath) {
        printState(node.state);
    }
}

function printState(state) {
    for (const row of state) {
        console.log(row.join(" "));
    }
    console.log();
}

main();
